### ISSUE TITLE
Cockpit terminal is nearly unusable for use on mobile phones

### ISSUE THREAD

**Issue by EPXOL** [cockpit-project/cockpit#18138]
I was using cockpit for a bit and I'm still using it but only thing that bothers me and thing that makes me not use cockpit as much as I would want is extremely bad experience with terminal on mobile. First idea of starting using cockpit was to close ssh and use only cockpit but after I found out that when I want to delete 1 character I must delete the whole word once then I can remove the one character i started using it less and less. This wasn't too big of a problem normally but when u have a big command and need to edit something at the start I usually just copied it to some text editor. But fixing this would make a lot better experience to use cockpit on daily basis.

**Comment by martinpitt** [comment]
Sorry for this bad experience! However, I'm afraid I need some more details. I tested this with Cockpit 282 running on Fedora and Firefox/Fennec as a client on Android 7, and  backspace on the on-screen keyboard deletes the last character just fine, not the whole word. What cockpit/mobile OS/version are you using? And which keyboard app? To be honest, deleting a whole word sounds more like something the keyboard app would do. Are you using the on-screen keyboard or a proper hardware one (bluetooth)?

That being said, the Terminal page *is* a frustrating experience for a lot of different reasons: the toolbar takes too much screen space, it's not possible to move the cursor or scroll through history (no exposed cursor keys), keyboard input uses auto-capitalization/correction, etc. Proper terminals like Termux or ConnectBot are much much better. Maybe there's at least a way to disable auto-correction, but even then IMHO it's too hard to use it for anything serious. It really at least needs a bluetooth keyboard.

**Comment by EPXOL** [comment]
I am using cockpit version 282 on Ubuntu 22.04.1 TLS machine. My mobile runs Android 11 and  I am using gboard as my default keyboard. I added a video of what I meen.

**Comment by EPXOL** [comment]
NOTE: This behaviour can be seen in podman containers in console too. I think it s smth with how cockpit handeles keypresses.

**Comment by EPXOL** [comment]
@martinpitt I tested the behaivoir with Microsoft keyboard and foud even worse resoults It copied the command like 3x before starting to remove it. I cannot seem to find the beyboard you wanted me to test. I can confirm that simular behavior si found on the site you provided but not exactly. The text I wrote was visible until I pressed spacebar. It didnt actually show up but when I tried to delete it it didnt seem to duplicate it. So there is actually a another bahaiviour between the site you provided and cockpit.

**Comment by EPXOL** [comment]
It avtually seems to me that xterm cannot handle single backspace press but it handles hold of backspace. Or the initial press of hold is sending some another key cousing it to duplicate. And maybe microsoft board is actually spamming the backspace on hold and that is why it duped more.

**Comment by KKoukiou** [comment]
Yes I can reproduce on my phone with gboard. https://jsfiddle.net/8vb7qken/ is completely broken on mobile, and even if I choose to view the desktop site on my mobile, it still unusable.

**Comment by EPXOL** [comment]
Mybe if instead of having a event lisener for keypresses you could make an invisible text area at the top of it. The in intervals check for changes and write then to background. backspace could just be handeled by like 5 spaces in the text area and in each loop cycle it would chceck how many spaces it starts woth and sould remove mumber of missing spaces back or send a key even to the server as a backspace. the thextarea sould shift back to left so it aligns with the text aready wrotten. This would make the input compatible with really any keyboard.

**Comment by EPXOL** [comment]
Any updates on this bug ?

**Comment by garrett** [comment]
Yep, as linked here on this issue, there are a few fixes that made it into the most recent version of Cockpit.

It's still not _optimal_ (we want to work further on this — you cannot enter control codes, for example), but at least it's possible to enter text and not have the menu pop up due to a recent regression.

Update to Cockpit 284 and it'll be improved for you. You should be able to do basic tasks in a terminal from a phone again.

**Comment by EPXOL** [comment]
Em btw how do u update from github source bcs the update didnt yet make it to my distro's repozitory.

**Comment by garrett** [comment]
@EPXOL: https://cockpit-project.org/running.html has information on how to install and update. Which distribution? Both Debian and Ubuntu have backports for newer versions of Cockpit. Several others, like Fedora, get a new version automatically.

**Comment by EPXOL** [comment]
I have Ubuntu on my server and the newest version I can install from backports is 283.

**Comment by martinpitt** [comment]
I uploaded 284 to backports this morning, but it failed. Presumably because I recently updated my GPG key, I need to sort that out first, sorry.

**Comment by martinpitt** [comment]
@EPXOL 284 is now [available in Ubuntu backports](https://launchpad.net/ubuntu/+source/cockpit).

**Comment by EPXOL** [comment]
Just upgraded it, right away noticed very nice improvement hitting ctrl+s doesn't open prompt to save the site, very neat. But the original issue seems to be the same, nothing changed for me, still same behaviour as before.

**Comment by EPXOL** [comment]
Just found an isssue at podman page gonna make a separate report for that.

**Comment by EPXOL** [comment]
Any updates yet ?

**Comment by KKoukiou** [comment]
@EPXOL sorry we did not have time to debug that more. What I know is that when I stopped using gboard on my android device it worked just fine. It also works without problems on the Iphone. All in all, the issue is xterm<->gboard interaction.

**Comment by romner-set** [comment]
I'm experiencing a similar issue on the Vivaldi browser (the snapshot version specifically), any letters typed don't actually get written into the terminal at all. Firefox works completely fine though, so I'm pretty sure it's a problem with Vivaldi specifically.

**Comment by Handrail9** [comment]
Hi, I'm writing to say I'm still facing this issue on Android 16 utilizing [Heliboard.](https://github.com/Helium314/HeliBoard) (Specifically, I'm facing frequent auto-corrects that do not "correct" until after I've hit enter, and I'm seeing the backspace duplicating words as I'm backspacing) The easiest solution has been to have a 2nd keyboard installed called [Hacker's Keyboard](https://github.com/klausw/hackerskeyboard), however they have not updated that keyboard since 2018, and if will presumably stop working soon. Would it be a good idea to open an issue over on Heliboard's side, or is this a Cockpit Terminal specific issue?
